src/Symfony/Component/Translation/Dumper/AbstractTreeFileDumper.php

```diff
@@ -0,0 +1,135 @@
+<?php
+
+/*
+ * This file is part of the Symfony package.
+ *
+ * (c) Fabien Potencier <fabien@symfony.com>
+ *
+ * For the full copyright and license information, please view the LICENSE
+ * file that was distributed with this source code.
+ */
+
+namespace Symfony\Component\Translation\Dumper;
+
+use Symfony\Component\Translation\MessageCatalogue;
+
+/**
+ * AbstractTreeFileDumper generates tree like structure from a message catalogue.
+ * e.g. this
+ *   'foo.bar1' => 'test1',
+ *   'foo.bar2' => 'test2'
+ * dumps as follows:
+ *   foo:
+ *     bar1: test1
+ *     bar2: test2
+ *
+ * @author Gennady Telegin <gtelegin@gmail.com>
+ */
+abstract class AbstractTreeFileDumper extends FileDumper
```

Comment:
The logic turning a MessageCatalogue into a nested array for the tree syntax is not specific to file dumpers IMO (you could imagine reusing it in a different context).

I suggest creating a dedicated class with the logic of your ``expandToTree`` method. The YamlFileDumper would then use this object as a collaborator to apply the logic.
The architecture will be much cleaner than relying on inheritance to bring the feature with a bunch of parent abstract classes.